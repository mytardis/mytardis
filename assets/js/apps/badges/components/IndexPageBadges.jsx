import React, { Fragment, useState } from 'react';
import ReactDOM from 'react-dom';

import PropTypes from 'prop-types';
import { css } from '@emotion/core';
import ExperimentLastUpdatedBadge from './ExperimentLastUpdateBadge';
import PublicAccessBadge from './PublicAccessBadge';
import DatasetCountBadge from './DatasetCountBadge';
import DatafileCountBadge from './DatafileCountBadge';
import { fetchExperimentData } from './utils/FetchData';
import Spinner from './utils/Spinner';


const IndexPageBadges = ({ experimentID }) => {
  const [expData, setExpData] = useState('');
  const [isLoading, setIsLoading] = useState(true);
  const spinnerCss = css`
    margin: auto;
    width: 20%;
    float: right;
    color: 9B9B9B;
  `;
  React.useEffect(() => {
    fetchExperimentData(experimentID).then((data) => {
      setExpData(data);
      setIsLoading(false);
    });
  }, []);
  return (
    isLoading ? <Spinner override={spinnerCss} />
      : (
        <Fragment>
          <ul className="list-inline float-right list-unstyled">
            <li className="mr-1 list-inline-item">
              <ExperimentLastUpdatedBadge experimentData={expData} />
            </li>
            <li className="mr-1 list-inline-item">
              <DatasetCountBadge experimentData={expData} />
            </li>
            <li className="mr-1 list-inline-item">
              <DatafileCountBadge experimentData={expData} />
            </li>
            <li className="mr-1 list-inline-item">
              <PublicAccessBadge experimentData={expData} />
            </li>
          </ul>
        </Fragment>
      )
  );
};

IndexPageBadges.propTypes = {
  experimentID: PropTypes.string.isRequired,
};

document.querySelectorAll('.badges')
  .forEach((domContainer) => {
    const experimentID = domContainer.id.split('-')[1];
    ReactDOM.render(
      <IndexPageBadges experimentID={experimentID} />, domContainer,
    );
  });

export default IndexPageBadges;
