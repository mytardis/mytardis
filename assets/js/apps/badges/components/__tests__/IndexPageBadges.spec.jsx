import React from 'react';
import ReactDOM from 'react-dom';
import { act } from '@testing-library/react';
import IndexPageBadges from '../IndexPageBadges';

global.fetch = require('jest-fetch-mock');

let fakeExperimentData = {
  update_time: '2020-01-13T14:00:08.908600',
  dataset_count: 1,
  datafile_count: 100,
  public_access: 1,
};
let container = null;
beforeEach(() => {
  jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
    json: () => Promise.resolve(fakeExperimentData),
  }));
  container = document.createElement('div');
  document.body.appendChild(container);
});
afterEach(() => {
  // cleanup on exiting
  container.remove();
  container = null;
});
describe('renders badges on index page', () => {
  it('should render correctly', async () => {
    await act(async () => {
      ReactDOM.render(<IndexPageBadges experimentID="123" />, container);
    });
    expect(container).toMatchSnapshot();
  });
  it('should render last updated badge', async () => {
    await act(async () => {
      ReactDOM.render(<IndexPageBadges experimentID="123" />, container);
    });
    expect(container.textContent).toContain('13th January 2020');
    expect(container.querySelector('span').attributes.getNamedItem('title').value)
      .toEqual('Last updated: Mon, Jan 13, 2020 2:00 PM');
  });
  it('should render dataset count badge', async () => {
    await act(async () => {
      ReactDOM.render(<IndexPageBadges experimentID="123" />, container);
    });
    expect(container.querySelector('ul').childNodes[1].firstChild.textContent.trim())
      .toEqual('1');
    expect(container.querySelector('ul').childNodes[1].firstChild.title).toEqual('1 dataset');
  });
  it('should render datafile count badge', async () => {
    await act(async () => {
      ReactDOM.render(<IndexPageBadges experimentID="123" />, container);
    });
    expect(container.querySelector('ul').childNodes[2].firstChild.textContent.trim())
      .toEqual('100');
    expect(container.querySelector('ul').childNodes[2].firstChild.title).toEqual('100 files');
  });
  it('should render Public access badge', async () => {
    await act(async () => {
      ReactDOM.render(<IndexPageBadges experimentID="123" />, container);
    });
    expect(container.querySelector('ul').childNodes[3].firstChild.textContent.trim())
      .toEqual('Private');
    expect(container.querySelector('ul').childNodes[3].firstChild.title.trim())
      .toEqual('No public access');
  });
  // remove the mock to ensure tests are completely isolated
  global.fetch.mockRestore();
});

describe('renders public access badge', () => {
  it('should render badge as private', async () => {
    await act(async () => {
      ReactDOM.render(<IndexPageBadges experimentID="123" />, container);
    });
    expect(container.querySelector('ul').childNodes[3].firstChild.title.trim())
      .toEqual('No public access');
  });
  it('should render badge as public', async () => {
    fakeExperimentData = {
      update_time: '2020-01-13T14:00:08.908600',
      dataset_count: 1,
      datafile_count: 100,
      public_access: 100,
    };
    await act(async () => {
      ReactDOM.render(<IndexPageBadges experimentID="123" />, container);
    });
    expect(container.querySelector('ul').childNodes[3].firstChild.title.trim())
      .toEqual('All data is public');
  });
  it('should render badge as Metadata', async () => {
    fakeExperimentData = {
      update_time: '2020-01-13T14:00:08.908600',
      dataset_count: 1,
      datafile_count: 100,
      public_access: 50,
    };
    await act(async () => {
      ReactDOM.render(<IndexPageBadges experimentID="123" />, container);
    });
    expect(container.querySelector('ul').childNodes[3].firstChild.title.trim())
      .toEqual('Only descriptions are public, not data');
  });
  it('should render badge as [PUBLICATION] Awaiting release', async () => {
    fakeExperimentData = {
      update_time: '2020-01-13T14:00:08.908600',
      dataset_count: 1,
      datafile_count: 100,
      public_access: 25,
    };
    await act(async () => {
      ReactDOM.render(<IndexPageBadges experimentID="123" />, container);
    });
    expect(container.querySelector('ul').childNodes[3].firstChild.title.trim())
      .toEqual('Under embargo and awaiting release');
  });
  global.fetch.mockRestore();
});
