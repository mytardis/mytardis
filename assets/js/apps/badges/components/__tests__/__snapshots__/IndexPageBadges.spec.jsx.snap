// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders badges on index page should render correctly 1`] = `
<div>
  <ul
    class="list-inline float-right list-unstyled"
  >
    <li
      class="mr-1 list-inline-item"
    >
      <span
        class="badge badge-info"
        title="Last updated: Mon, Jan 13, 2020 2:00 PM"
      >
        <i
          class="fa fa-clock-o"
        />
         
        13th January 2020
      </span>
    </li>
    <li
      class="mr-1 list-inline-item"
    >
      <span
        class="badge badge-info"
        title="1 dataset"
      >
        <i
          class="fa fa-folder"
        />
         
        1
      </span>
    </li>
    <li
      class="mr-1 list-inline-item"
    >
      <span
        class="badge badge-info"
        title="100 files"
      >
        <i
          class="fa fa-file"
        />
         
        100
      </span>
    </li>
    <li
      class="mr-1 list-inline-item"
    >
      <span
        class="badge badge-secondary"
        content=" No public access"
        title=" No public access"
      >
        <i
          class="fa fa-eye-slash"
        />
         Private
      </span>
    </li>
  </ul>
</div>
`;
